packages/git/src/browser/git-base-widget.ts

```diff
@@ -5,18 +5,68 @@
  * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
  */
 
-import { VirtualWidget } from "@theia/core/lib/browser";
+import { VirtualWidget, SELECTED_CLASS } from "@theia/core/lib/browser";
 import { GitFileStatus, Repository, GitFileChange } from '../common';
 import URI from "@theia/core/lib/common/uri";
 import { GitRepositoryProvider } from "./git-repository-provider";
 import { LabelProvider } from "@theia/core/lib/browser/label-provider";
+import { Message } from "@phosphor/messaging";
+import { Key } from "@theia/core/lib/browser/keys";
+import { SelectionService } from "@theia/core";
+import { ElementExt } from "@phosphor/domutils";
 
-export class GitBaseWidget extends VirtualWidget {
+export enum SelectDirection {
+    NEXT, PREVIOUS
+}
+
+export interface GitFileChangeNode extends GitFileChange {
+    readonly icon: string;
+    readonly label: string;
+    readonly description: string;
+    readonly caption?: string;
+    readonly extraIconClassName?: string;
+    readonly commitSha?: string;
+    selected?: boolean;
+}
+
+export namespace GitFileChangeNode {
+    export function is(node: any): node is GitFileChangeNode {
+        return 'uri' in node && 'status' in node && 'description' in node && 'label' in node && 'icon' in node;
+    }
+}
+
+export abstract class GitBaseWidget<T extends { selected?: boolean }> extends VirtualWidget {
+
+    protected gitNodes: T[];
+    protected abstract readonly scrollContainer: string;
 
     constructor(
         protected readonly repositoryProvider: GitRepositoryProvider,
-        protected readonly labelProvider: LabelProvider) {
+        protected readonly labelProvider: LabelProvider,
+        protected readonly selectionService: SelectionService) {
         super();
+
+        selectionService.onSelectionChanged((c: T) => {
+            c.selected = true;
+            this.update();
+        });
+
+        this.node.tabIndex = 0;
+    }
+
+    protected onActivateRequest(msg: Message): void {
+        super.onActivateRequest(msg);
+        this.node.focus();
+    }
+
+    protected onUpdateRequest(msg: Message): void {
+        super.onUpdateRequest(msg);
+
+        const selected = this.node.getElementsByClassName(SELECTED_CLASS)[0];
+        const scrollArea = document.getElementById(this.scrollContainer);
```

Comment:
then please make sure, that `scrollContainer` is unique. `GitHistoryWidget` seems to be bound in transient scope, which means, in theory more than one instance is possible.

Reply:
well, this is very convenient that the widget manager will return you the very same instance of a widget, but only if the internal key (a combination of id and options) is the same. which means, if I would request a git widget with different options, I'd get a new instance in this case. I'm not planing to do so, but the point is, the selection of the scroll container should be a bit more robust here. you could just add a counter or use a timestamp to salt it.